File: src/commonTest/kotlin/com/scoreo/application/GetGameTypesUseCaseTest.kt
package com.scoreo.application

import com.scoreo.infrastructure.InMemoryGameTypeRepository
import com.scoreo.domain.model.GameType
import com.scoreo.domain.model.WinCondition
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.test.assertNull

class GetGameTypesUseCaseTest {

    @Test
    fun `returns empty list when no game types saved`() {
        val repo = InMemoryGameTypeRepository()
        val useCase = GetGameTypesUseCase(repo)
        assertTrue(useCase().isEmpty())
    }

    @Test
    fun `returns all saved game types`() {
        val repo = InMemoryGameTypeRepository()
        repo.save(GameType("1", "Belote", WinCondition.HIGHEST_SCORE))
        repo.save(GameType("2", "Golf", WinCondition.LOWEST_SCORE))
        val useCase = GetGameTypesUseCase(repo)
        val result = useCase()
        assertEquals(2, result.size)
        assertEquals(listOf("Belote", "Golf"), result.map { it.name })
    }

    @Test
    fun `excludes inactive game types by default`() {
        val repo = InMemoryGameTypeRepository()
        repo.save(GameType("1", "Belote", WinCondition.HIGHEST_SCORE, active = true))
        repo.save(GameType("2", "Golf", WinCondition.LOWEST_SCORE, active = false))
        repo.save(GameType("3", "Bridge", WinCondition.HIGHEST_SCORE, active = true))
        val useCase = GetGameTypesUseCase(repo)
        val result = useCase()
        assertEquals(2, result.size)
        assertEquals(listOf("Belote", "Bridge"), result.map { it.name })
    }

    @Test
    fun `findById includes inactive games`() {
        val repo = InMemoryGameTypeRepository()
        val archivedGame = GameType("1", "Archived", WinCondition.HIGHEST_SCORE, active = false)
        repo.save(archivedGame)
        
        val found = repo.findById("1")
        assertEquals("Archived", found?.name)
        assertEquals(false, found?.active)
    }

    @Test
    fun `getAll with includeInactive true returns all games`() {
        val repo = InMemoryGameTypeRepository()
        repo.save(GameType("1", "Belote", WinCondition.HIGHEST_SCORE, active = true))
        repo.save(GameType("2", "Golf", WinCondition.LOWEST_SCORE, active = false))
        repo.save(GameType("3", "Bridge", WinCondition.HIGHEST_SCORE, active = true))
        val result = repo.getAll(includeInactive = true)
        assertEquals(3, result.size)
        assertEquals(listOf("Belote", "Golf", "Bridge"), result.map { it.name })
    }
}

File: apps/scoreo/src/application/getGameTypesUseCase.test.ts
import { describe, expect, it } from 'vitest'
import { InMemoryGameTypeRepository } from '../infrastructure/testing/inMemoryGameTypeRepository'
import { GetGameTypesUseCase } from './getGameTypesUseCase'

function gt(id: string, name: string, winCondition: 'HIGHEST_SCORE' | 'LOWEST_SCORE', active = true) {
  return {
    id,
    name,
    winCondition,
    tieBreakRule: 'NONE' as const,
    tieBreakCondition: 'HIGHEST_SCORE' as const,
    tieBreakLabel: null,
    moduleId: null,
    active,
  }
}

describe('GetGameTypesUseCase', () => {
  it('returns empty list when no game types saved', () => {
    const repo = new InMemoryGameTypeRepository()
    const useCase = new GetGameTypesUseCase(repo)
    expect(useCase.invoke()).toHaveLength(0)
  })

  it('returns all saved game types', () => {
    const repo = new InMemoryGameTypeRepository()
    repo.save(gt('1', 'Belote', 'HIGHEST_SCORE'))
    repo.save(gt('2', 'Golf', 'LOWEST_SCORE'))
    const useCase = new GetGameTypesUseCase(repo)
    const result = useCase.invoke()
    expect(result).toHaveLength(2)
    expect(result.map((g) => g.name)).toEqual(['Belote', 'Golf'])
  })

  it('excludes inactive game types by default', () => {
    const repo = new InMemoryGameTypeRepository()
    repo.save(gt('1', 'Belote', 'HIGHEST_SCORE', true))
    repo.save(gt('2', 'Golf', 'LOWEST_SCORE', false))
    repo.save(gt('3', 'Bridge', 'HIGHEST_SCORE', true))
    const useCase = new GetGameTypesUseCase(repo)
    const result = useCase.invoke()
    expect(result).toHaveLength(2)
    expect(result.map((g) => g.name)).toEqual(['Belote', 'Bridge'])
  })

  it('findById includes inactive games', () => {
    const repo = new InMemoryGameTypeRepository()
    repo.save(gt('1', 'Archived', 'HIGHEST_SCORE', false))

    const found = repo.findById('1')
    expect(found?.name).toBe('Archived')
    expect(found?.active).toBe(false)
  })

  it('getAll with includeInactive true returns all games', () => {
    const repo = new InMemoryGameTypeRepository()
    repo.save(gt('1', 'Belote', 'HIGHEST_SCORE', true))
    repo.save(gt('2', 'Golf', 'LOWEST_SCORE', false))
    repo.save(gt('3', 'Bridge', 'HIGHEST_SCORE', true))
    const result = repo.getAll(true)
    expect(result).toHaveLength(3)
    expect(result.map((g) => g.name)).toEqual(['Belote', 'Golf', 'Bridge'])
  })
})
